### ISSUE TITLE
Windows: GC thread crash

### ISSUE THREAD

**Issue by p0nce** [ldc-developers/ldc#1298]
I'm on the verge on having working Windows builds made with LDC 0.17.0, for both 32-bit and 64-bit. Performance is 2x higher and the sun is shining.

However, I have to disable the GC in my shared libraries which is breaking my heart.
I'm pretty sure I've seen the GC collect objects while I was in their constructor. Another time when attaching a debugger, the GC thread crash in marking phase, and let all the other thread paused.

For some odd reason this isn't a problem when a D program is loading the shared libraries, but it is when it's a C++ program.

This happens regardless of if the VS runtime is statically or dynamically linked.

Is this a known problem?

**Comment by kinke** [comment]
I don't use D at all except when playing with LDC, so I don't know anything about real-world problems. ;)
It sounds like an issue with static constructors inside the D modules...

**Comment by JohanEngelen** [comment]
In a C++ program, linking to a D lib, you will have to call `rt_init();` of the D lib to initialize D's runtime.

**Comment by p0nce** [comment]
That's what I'm doing. The same program works without GC crash with all DMD versions since 2.067.

**Comment by p0nce** [comment]
Actually, I must check that the DLL WinMain is called by LoadLibrary, could be that.

**Comment by kinke** [comment]
We've had that issue with static libraries a while ago, that was due to a bad TLS range iirc; see https://github.com/ldc-developers/druntime/blob/ldc/src/ldc/msvc.c. `_data_start/end__` or `_bss_start/end__` were wrong back then, which led the GC to collect everything on each cycle. Could you please check these values?

**Comment by kinke** [comment]
I've just seen that we use `GetModuleHandle(NULL)` in that druntime file. According to MSDN, this yields the handle to the .exe module. In your case, that will be the C++ .exe. We probably need the module which druntime has been linked into...

**Comment by p0nce** [comment]
> Could you please check these values?
Not sure what you mean.

I can perhaps work towards a minimal sample it should bea easy to reproduce with many allocations.

And now that I see your message, I also use GetModuleHandle(NULL) in places.

**Comment by p0nce** [comment]
I have something reproducible with  ldc2-0.17.0-win64-msvc.

**OS:** Windows 10

**Source:**

```d
import core.sys.windows.windows;
import core.sys.windows.dll;
import core.stdc.stdio;
import std.stdio;

extern (Windows) BOOL DllMain(HINSTANCE hInstance, ULONG ulReason, LPVOID pvReserved)
{
    switch (ulReason)
    {
        case DLL_PROCESS_ATTACH:
            dll_process_attach(hInstance, true);
            printf("DLL_PROCESS_ATTACH\n");
            break;

        case DLL_PROCESS_DETACH:
            printf("DLL_PROCESS_DETACH\n");
            dll_process_detach(hInstance, true);
            break;

        case DLL_THREAD_ATTACH:
            dll_thread_attach(false, true);
            printf("DLL_THREAD_ATTACH\n");
            break;

        case DLL_THREAD_DETACH:
            printf("DLL_THREAD_DETACH\n");
            dll_thread_attach(true, true);
            break;

        default:
            break;
    }
    return true;
}

extern(C) nothrow void* VSTPluginMain(void* hostCallback)
{
    try
    {
        thread_attachThis();

        // this loop is for provoking a GC collect
        A[] a;
        for (int i = 0; i < 1000; ++i)
        {
            a ~= new B();
        }

        return null;
    }
    catch(Exception e)
    {
        return null;
    }
}

class A 
{
}

class B : A
{
    ~this()
    {
      import std.stdio;
      writeln("wtf");
    }
}
```

This is a fake VST which return NULL hence unloadable. But it should not crash hosts.

**To build:**
`ldc2 -shared -ofreproldc.dll -d-debug -g -w -oq -I. reproldc.d -L/def:module.def -shared -g -relocation-model=pic`

**To reproduce**

- Get a 64-bit VST host like eg. http://www.hermannseib.com/programs/vsthostx64.zip

- Load the shared library using the menu: File / New plugin then select `reproldc.dll`

Success is defined by not crashing the host. The bug is that we have a high probability to crash in `VSTPluginMain`.

Will get back to you with a backtrace.

**Comment by p0nce** [comment]
![stacktrace](https://cloud.githubusercontent.com/assets/1067485/13086551/f7bf2f92-d4e3-11e5-8d96-f51ee1cac03f.png)

What Windbg says.

**Comment by kinke** [comment]
See https://github.com/ldc-developers/druntime/pull/61. I haven't had the time to test it properly yet.

**Comment by kinke** [comment]
That fix still crashes the host, identical callstack. :/

**Comment by p0nce** [comment]
Would someone take bounties for this bug? How much?

**Comment by rainers** [comment]
> extern(C) nothrow void* VSTPluginMain(void* hostCallback) { try { thread_attachThis();

Not sure it will help, but calling `thread_attachThis();` here is wrong, it should be done only once per thread by passing `true` as the first argument to `dll_thread_attach`.

**Comment by p0nce** [comment]
The problem is that I need the audio thread detached, so if I attach all the threads with `dll_thread_attach` I have to call `thread_detachThis()` in the audio thread and it's not `@nogc`. This hasn't caused a problem AFAIK but maybe I should reevaluate.

**Comment by p0nce** [comment]
Tested that but it still crash with an `InvalidMemoryOperationError`
=> modified the test above, also included `module.def` that I had forgotten.

**Comment by rainers** [comment]
> he problem is that I need the audio thread detached, so if I attach all the threads with dll_thread_attach in DllMain I have to call thread_detachThis() in the audio thread and it's not marked @nogc.

Shortly after having written my comment and having left, I remembered your use case. (OT: My VST2-fu is a bit rusty, but is the callback never called from the audio thread? Wouldn't that depend on the host, especially when setting parameters?)
Using `thread_attachThis()` should work as you did because there is a check in it whether the thread is already attached. But it involves GC locking and I would avoid calling it by checking a thread local.

I hope I can look into this the next couple of days, as I want to try building Visual D with LDC, too.

**Comment by p0nce** [comment]
> But it involves GC locking and I would avoid calling it by checking a thread local.

This is unrelated to the issue at hand, I think you mean something like:
```d
bool thread_attached = false; // at top-level
```
```d
if (!thread_attached)
{
    thread_attachThis();
    thread_attached = true;
}
```

But I'm not sure if the thread-local actually exist if the thread is not attached yet. I don't understand TLS well.

**Comment by p0nce** [comment]
> (OT: My VST2-fu is a bit rusty, but is the callback never called from the audio thread? Wouldn't that depend on the host, especially when setting parameters?)

There is several callbacks in VST, nothing is said about which threads call what. So I attach threads as needed depending on the callback they call. If the audio thread ends up attached, it means that the host didn't care and didn't have a thread dedicated for it.

The example was mixed up for bad reasons, happened while reducing.

**Comment by rainers** [comment]
> But I'm not sure if the thread-local actually exist if the thread is not attached yet.

TLS is created by the OS, so it's already there (for XP, the code in DllMain takes care of this). So your code looks as I meant it.

I can reproduce the issue and have debugged it a bit. There seems to be something bad about capturing the stack top/bottom of suspended threads.

BTW:  `dll_process_attach(hInstance, true);` in `case DLL_PROCESS_ATTACH:` also attaches to all existing threads, including audio threads that are already running. You can pass `false` as second argument to avoid that. When doing so, the crashes go away for me.

**Comment by p0nce** [comment]
> BTW: dll_process_attach(hInstance, true); in case DLL_PROCESS_ATTACH: also attaches to all existing threads, including audio threads that are already running. You can pass false as second argument to avoid that. When doing so, the crashes go away for me.

This looks like a big mistake on my part. And I thought that was figured out. Generally speaking, I can't thank you enough for your work!

**Comment by rainers** [comment]
The crashes might reappear when more threads are created and attached to later. Also the default version that attaches all threads should not crash (but only cause clicks in the audio if the GC suspends these threads). I'll try to figure it out tomorrow...

**Comment by rainers** [comment]
It think I figured it out: https://github.com/rainers/druntime/commit/464f7092a943ef1dccc6db13cb61bf618699d840
I don't know why it works/seems to work with dmd, though.

**Comment by p0nce** [comment]
Hi there, still concerned that I have to disable the GC (though it doesn't seem to cause too much problems). Would be cool to have the fix in 1.0.0
If you need some sort of compensation feel free to ask away!

**Comment by rainers** [comment]
Did you try the patch rainers/druntime@464f709 above? I wasn't sure why this is not an issue with dmd, so I did not yet post a pull request to ldc/dlang.

**Comment by rainers** [comment]
I just browsed the history a bit, and it seems use of TlsAlloc has been removed from druntime in 2013, but it's still on ldc/ltsmaster. The issue should not happen with master.

**Comment by p0nce** [comment]
> Did you try the patch rainers/druntime@464f709 above? I wasn't sure why this is not an issue with dmd, so I did not yet post a pull request to ldc/dlang.

No, I didn't. Is this required? If you did see the problem disappear, I'll trust that for sure.

> I just browsed the history a bit, and it seems use of TlsAlloc has been removed from druntime in 2013, but it's still on ldc/ltsmaster. The issue should not happen with master.

Thanks. So in layman terms we only need to wait?

**Comment by rainers** [comment]
> No, I didn't. Is this required? If you did see the problem disappear, I'll trust that for sure.

It would be nice to know whether it fixes the issues for you, too. It's not required, though, as I understand that it's quit bothersome to compile LDC on windows.

> Thanks. So in layman terms we only need to wait?

Could work;-) I'm not quite sure what branch is used to build 1.0.0. @redstar ?

I'll post a PR to druntime upstream, as this is still a real problem for DLLs that use TlsAlloc.

**Comment by redstar** [comment]
I use the release-1.0.0 branch for the release build. This is (currently) one commit behind master.
@p0nce I published the beta1 version this morning. Maybe you can try it?

**Comment by p0nce** [comment]
Sure I'll be able to try it tomorrow if all goes well.

**Comment by p0nce** [comment]
I can attest this is fixed as of LDC 1.0.0-b1
Not able to reproduce the crash in the test case and in larger projects.
More over my products work way better now that the GC is enabled back.
Speed-wise, performance seems unchanged vs LDC 0.17.0-b1
